node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/servicedump/VespaServiceDumperImpl.java

```diff
@@ -0,0 +1,163 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.node.admin.maintenance.servicedump;
+
+import com.yahoo.vespa.hosted.node.admin.configserver.noderepository.NodeAttributes;
+import com.yahoo.vespa.hosted.node.admin.configserver.noderepository.NodeRepository;
+import com.yahoo.vespa.hosted.node.admin.configserver.noderepository.NodeSpec;
+import com.yahoo.vespa.hosted.node.admin.container.ContainerOperations;
+import com.yahoo.vespa.hosted.node.admin.maintenance.sync.SyncClient;
+import com.yahoo.vespa.hosted.node.admin.maintenance.sync.SyncFileInfo;
+import com.yahoo.vespa.hosted.node.admin.nodeagent.NodeAgentContext;
+import com.yahoo.vespa.hosted.node.admin.task.util.file.FileFinder;
+import com.yahoo.vespa.hosted.node.admin.task.util.process.CommandResult;
+
+import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * Generates dumps for Vespa services and uploads resulting files to S3.
+ *
+ * @author bjorncs
+ */
+public class VespaServiceDumperImpl implements VespaServiceDumper {
+
+    private static final Logger log = Logger.getLogger(VespaServiceDumperImpl.class.getName());
+    private static final String DIRECTORY_IN_NODE = "/opt/vespa/tmp/vespa-service-dump";
+
+    private final ContainerOperations container;
+    private final SyncClient syncClient;
+    private final NodeRepository nodeRepository;
+
+    public VespaServiceDumperImpl(ContainerOperations container, SyncClient syncClient, NodeRepository nodeRepository) {
+        this.container = container;
+        this.syncClient = syncClient;
+        this.nodeRepository = nodeRepository;
+    }
+
+    @Override
+    public void processServiceDumpRequest(NodeAgentContext context) {
+        Instant startedAt = Instant.now();
+        NodeSpec nodeSpec = context.node();
+        ServiceDumpReport request = nodeSpec.reports().getReport(ServiceDumpReport.REPORT_ID, ServiceDumpReport.class)
+                .orElse(null);
+        if (request == null || request.failedAt() != null || request.completedAt() != null) {
```

Comment:
Add a method on ServiceDumpReport that returns true if it has reached a terminal state?